lib/rake_github/task_sets/repository.rb

```diff
@@ -37,6 +37,7 @@ class Repository < RakeFactory::TaskSet
             ts.deploy_keys_destroy_task_name
           }
       task Tasks::PullRequests::Merge,
+           # TODO: should this be here?
           argument_names: [:branch_name, :commit_message],
```

Comment:
It should make sense to instantiate the `Tasks::PullRequests::Merge` task directly since task sets are really just a convenience for a collection of related tasks. By putting the argument names here, they aren't available in direct instantiations.

It looks to me like `:branch_name` and `:commit_message` should be default arguments of the task since it's likely the user will always want to specify them at task invocation time rather than task definition time. However, then they can't easily be parameters of the task since `rake_factory` doesn't have much facility for using arg values as the values for parameters when they are provided. It would be nice to keep them as parameters so that validations and defaults kick in though. Potentially, this is something we could add to `rake_factory` but need to think it through.

There's an example here: https://github.com/infrablocks/rake_easy_rsa/blob/master/lib/rake_easy_rsa/tasks/certificate/renew.rb of using arguments in the task itself. In this case, the argument names are always set on the task and can't be removed, however other argument names can be provided which prefix the default arguments. This is an approach we could use.

We could also specify both default argument names and parameters allowing both invocation and definition time configuration but would need to work out the precedence between them in that case.

@Gryff What do you think?

Reply:
@jonassvalin specifically this which needs some more thought.

Reply:
One alternative would be to define the default arguments on the task but also define the parameters. Then when they are used, do something like:

```ruby
branch_name = args.branch_name.strip.empty? ? t.branch_name : args.branch_name.strip
commit_message = args.commit_message.strip.empty? ? t.commit_message : args.commit_message.strip
```

This would give precedence to args but when missing, use whatever was configured as a parameter. The only downside would be that the args don't benefit from the type conversion of parameters.

Reply:
Another thing to consider here is that the repository task set is likely to grow pretty big, so parameter names declared on the task set should be quite specific to the task they are for unless they are very general.

So if we do expose these parameters on the repository task set, we should use names like `pull_request_merge_branch_name` and `pull_request_merge_commit_message`.